app/src/main/java/org/mozilla/fenix/library/history/HistoryAdapter.kt

```diff
@@ -8,122 +8,229 @@ import android.view.LayoutInflater
-                                it is PendingDeletionHistory.MetaData &&
-                                    it.key == historyMetadata.historyMetadataKey &&
-                                    it.visitedAt == historyMetadata.visitedAt
-                            } != null
-                        }.also {
-                            if (it == current.items.size) {
-                                isPendingDeletion = true
-                            }
+                holder.bind(item, mode, groupPendingDeletionCount)
+            }
+            is TimeGroupViewHolder -> holder.bind(item as HistoryViewItem.TimeGroupHeader)
+            is RecentlyClosedViewHolder -> holder.bind(item as HistoryViewItem.RecentlyClosedItem)
+            is SyncedHistoryViewHolder -> holder.bind(item as HistoryViewItem.SyncedHistoryItem)
+            is EmptyViewHolder -> {
+                holder.bind(item as HistoryViewItem.EmptyHistoryItem)
+                // Resize emptyViewHolder to fill the rest of the recyclerview space.
+                val lastItemView = holder.itemView
+                lastItemView.viewTreeObserver.addOnGlobalLayoutListener {
+                    recycler?.height?.let { recyclerViewHeight ->
+                        val lastItemBottom = lastItemView.bottom
+                        val heightDifference = recyclerViewHeight - lastItemBottom
+                        if (heightDifference > 0) {
+                            lastItemView.layoutParams.height = lastItemView.height + heightDifference
+                            lastItemView.requestLayout()
                         }
                     }
                 }
-                else -> {}
             }
         }
+    }
 
-        // Add or remove the header and position to the map depending on it's deletion status
-        if (itemsWithHeaders.containsKey(current.historyTimeGroup)) {
-            if (isPendingDeletion && itemsWithHeaders[current.historyTimeGroup] == position) {
-                itemsWithHeaders.remove(current.historyTimeGroup)
-            } else if (isPendingDeletion && itemsWithHeaders[current.historyTimeGroup] != position) {
-                // do nothing
-            } else {
-                if (position <= itemsWithHeaders[current.historyTimeGroup] as Int) {
-                    itemsWithHeaders[current.historyTimeGroup] = position
-                    timeGroup = current.historyTimeGroup
+    private fun onDeleteClicked(adapterPosition: Int) {
+        // The click might have happened during animation.
+        if (adapterPosition == RecyclerView.NO_POSITION) return
+
+        getItem(adapterPosition)?.let {
+            when (it) {
+                is HistoryViewItem.HistoryItem -> it.data
+                is HistoryViewItem.HistoryGroupItem -> it.data
+                else -> null
+            }?.let { historyItem ->
+                val headerToRemove = calculateTimeGroupsToRemove(setOf(historyItem))
```

Comment:
I am a bit torn on that, wouldn't it actually be more confusing to the reader? We are removing just one history item, so the `calculateTimeGroupsToRemove` should have only one header to remove. Do we want to go with what's actually happening or what the return value of the function is?

Reply:
As a developer you know that this particular call you're only passing in one item in the set and it will only return one item.  However, anyone can update the function to return multiple items in the future.  Since it's possible (although currently not) we should stay consistent.